Translation of template.yml to CDK

nataliemona opened the pull request:
Not trying to do anything too complicated here. Not trying to change the functionality at all. Just translating exactly what we have in template.yml into CDK and removing some tasks from tasks.py that are no longer necessary.

nataliemona committed:
setting up cdk

nataliemona committed:
add codebuild, on demand endpoint, op counts function, on demand function, sns topic, loc notifications function, some alarms

nataliemona committed:
vpc, lambda env vars, some cloudwatch alarms, lambda permissions

nataliemona committed:
successful ingest (fixed IAM issue)

nataliemona committed:
cloudwatch alarms and metric filters

nataliemona committed:
remove now unecessary tasks

nataliemona committed:
remove unused import

nataliemona committed:
try to pass new_recording endpoint url to on demand function

nataliemona committed:
successfully give on demand function the WEBHOOK_ENDPOINT_URL

nataliemona committed:
give codebuild permissions to update and publish lambda

nataliemona committed:
add allow_multiple_ingests query param

lbjay commented on app.py:
```diff
@@ -0,0 +1,30 @@
+from dotenv import load_dotenv
+from os import getenv as env
+
+from zoom_recording_ingester_cdk.zoom_recording_ingester_cdk_stack \
```
this is kinda ugly. is this normal for python cdk projects? i'm used to the constructs being in a `./lib` directory. I know it's just aesthetics but would love for this to be something like `from cdk import ZoomRecordingIngesterCdkStack`. Rename the package directory and drop an `__init__.py` in there that imports the stack class?

lbjay commented on app.py:
```diff
@@ -0,0 +1,30 @@
+from dotenv import load_dotenv
+from os import getenv as env
+
+from zoom_recording_ingester_cdk.zoom_recording_ingester_cdk_stack \
```
this is kinda ugly. is this normal for python cdk projects? i'm used to the constructs being in a `./lib` directory. I know it's just aesthetics but would love for this to be something like `from cdk import ZoomRecordingIngesterCdkStack`. Rename the package directory and drop an `__init__.py` in there that imports the stack class?

nataliemona replied:
I didn't write this part, it's the default generated when you do 'cdk init' for python.
Here's the relevant page in the CDK python workshop https://cdkworkshop.com/30-python/20-create-project/300-structure.html

However, I definitely want design feedback. If we can restructure it without breaking anything than I'm for it.

lbjay commented on app.py:
```diff
@@ -0,0 +1,30 @@
+
+load_dotenv(join(dirname(__file__), '.env'))
+
+CDK_DEFAULT_ACCOUNT = env("CDK_DEFAULT_ACCOUNT")
+CDK_DEFAULT_REGION = env("CDK_DEFAULT_REGION")
```
these need to be in `.env`?

nataliemona replied:
Nope.
"To make the stack deployable to a different target, but to determine the target at synthesis time, your stack can use two environment variables provided by the AWS CDK CLI: CDK_DEFAULT_ACCOUNT and CDK_DEFAULT_REGION."
https://docs.aws.amazon.com/cdk/latest/guide/environments.html

lbjay commented on requirements.in:
```diff
@@ -13,6 +13,19 @@ requests-mock
+aws-cdk.aws_lambda
+aws-cdk.aws_dynamodb
+aws-cdk.aws_apigateway
+aws-cdk.aws_events_targets==1.32.0
```
why just these two pinned?

nataliemona replied:
Because if you pin nothing, there will be conflicts because some modules depend on older versions of other modules. So you sometimes need to pin the modules that are dependencies to older versions.

lbjay commented on tasks.py:
```diff
@@ -208,60 +207,6 @@ def release(ctx, function=None, description=None):
         __update_release_alias(ctx, func, new_version, description)
 
 
-@task
```
this is just deprecated?

nataliemona replied:
With the current number of recordings we have in our account these days it's so slow it's basically unusable and it would require making an API request repeatedly that is likely to disrupt other services. And we have much better ways now to get recordings to ingest or to find the uuid of a specific recording.

lbjay commented on zoom_recording_ingester_cdk/zoom_recording_ingester_cdk_stack.py:
```diff
@@ -0,0 +1,683 @@
+from aws_cdk import (
```
First comment off the bat is it's strange to me for everything to be in a single, huge Stack class. I tried to emulate the examples I was seeing and ended up with lots of separate classes when doing the caccl-deploy stuff: `https://github.com/harvard-edtech/caccl-deploy/tree/master/deploy/lib`.

nataliemona replied:
I agree. I just couldn't find an example anywhere of a python CDK app written with multiple classes.

lbjay commented on zoom_recording_ingester_cdk/zoom_recording_ingester_cdk_stack.py:
```diff
@@ -0,0 +1,683 @@
+        super().__init__(scope, id, **kwargs)
+
+        # Lambda alias that points to the "live" version of the function
+        self.lambda_release_alias = self._getenv("LAMBDA_RELEASE_ALIAS")
```
again this could be me not being used to the pythonic cdk but it seems like it would be better to collect all them environment vars in `app.py` and pass them into the Stack constructor. Explicitly define the params too, vs just using `**kwargs`. Could make for a very long constructor, but I still think it's better.

lbjay commented on zoom_recording_ingester_cdk/zoom_recording_ingester_cdk_stack.py:
```diff
@@ -0,0 +1,683 @@
+        )
+
+        vpc_id, sg_id = self.vpc_components()
+        self.oc_vpc = ec2.Vpc.from_lookup(self, "OcVpc", vpc_id=vpc_id)
```
I think this should also be done outside of the stack  class. If there's no vpc found things should fail/error before we get to this point IMO.

lbjay commented on zoom_recording_ingester_cdk/zoom_recording_ingester_cdk_stack.py:
```diff
@@ -0,0 +1,683 @@
+        SQS queues
+        """
+
+        downloads_queue, downloads_dlq = self._create_queue("downloader")
```
IMO this should be a separate queue construct class instead of a method

lbjay commented on zoom_recording_ingester_cdk/zoom_recording_ingester_cdk_stack.py:
```diff
@@ -0,0 +1,683 @@
+            "POST",
+            request_parameters={
+                "method.request.querystring.type": True,
+                "method.request.querystring.content": True
```
Is this `content` param actually relevant anymore. I actually wonder if any of these request params are still necessary.

nataliemona replied:
Ok. I was just blindly coping things over from the previous setup.

lbjay commented on zoom_recording_ingester_cdk/zoom_recording_ingester_cdk_stack.py:
```diff
@@ -0,0 +1,683 @@
+            "POST",
+            on_demand_integration,
+            request_parameters={
+                "method.request.querystring.uuid": True,
+                "method.request.querystring.oc_series_id": False,
+                "method.request.querystring.allow_multiple_ingests": False
```
same here. i don't remember what these even do. our methods are all POST so shouldn't have to care about querystring params.

nataliemona replied:
I don't know. I was just copying the template.yml. But ok!

lbjay commented on zoom_recording_ingester_cdk/zoom_recording_ingester_cdk_stack.py:
```diff
@@ -0,0 +1,683 @@
+            )
+        )
+
+        """
+        SNS Topic
+        """
```
I feel like everywhere you have one of these `"""` headers could be a new class/construct

lbjay committed:
removed the pinned versions and ran `pip-compile -U`; worked fine for me

lbjay committed:
refactoring of cdk stack & constructs

lbjay committed:
allow vars in `.env` to override existing, e.g. `AWS_PROFILE`

lbjay committed:
initial work on using cfn outputs to manage/share resource names

* consolidate getting aws account id
* add cfn outupts to stack status output

lbjay committed:
attache resource policy to api and enable CORS for ingest endpoint

lbjay committed:
set limited log retention for lambdas and api

lbjay committed:
use a shared module of resource name constants, `cdk/names.py`

also:

* increase build juice; use pip-sync to ensure the correct version combos
*`LAMBDA_RELEASE_ALIAS` is also now name constant instead of a `.env` setting
* more ckd package updates

lbjay commented on buildspec.yml:
```diff
@@ -4,14 +4,12 @@ phases:
   install:
     runtime-versions:
       python: 3.8
-    commands:
```
I saw in the build logs that none of these packages were actually installing. Most weren't even found, some, like zip/unzip were already the latest versions.

nataliemona replied:
Tested the `invoke codebuild --revision=cdk-devel` and it worked, so no complaints here

lbjay commented on cdk.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "app": "python3 -m cdk.app"
```
this is slightly odd, but doing `python3 cdk/app.py` gave me import errors. probably other ways to get around it, but using `-m` seems sufficiently pythonic.

nataliemona replied:
This explains the solution [https://napuzba.com/a/import-error-relative-no-parent](https://napuzba.com/a/import-error-relative-no-parent)... but I think `-m` is fine in this case?

lbjay commented on cdk/api.py:
```diff
@@ -0,0 +1,187 @@
+            ]
+        )
+
+        self.rest_api_name = f"{stack_name}-{names.REST_API}"
```
If you hate this naming mechanism we can discuss. There's one step further I could take it where we could have the names module injecting the stack name.

nataliemona replied:
I'm only concerned about naming if it impacts my ability to recreate a dashboard from another stack, find logs, read the stack diff, or find a resource related to the stack... so I have to run/test the CDK app to see what my preferences are.

lbjay commented on cdk/api.py:
```diff
@@ -0,0 +1,187 @@
+                    principals=[iam.AnyPrincipal()],
+                    # note that the policy is a prop of the api which cannot reference itself
+                    # see the Cloudformation documentation for api gateway policy attribute
+                    resources=[core.Fn.join('', ['execute-api:/', '*'])]
```
still kind of agog that this really works

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-apigateway-restapi.html#cfn-apigateway-restapi-policy

nataliemona replied:
Why do we need this? Is this the incorporation of steps that were previously manual steps?

lbjay replied:
yes

lbjay commented on cdk/api.py:
```diff
@@ -0,0 +1,187 @@
+            endpoint_url("new_recording")
+        )
+
+        core.CfnOutput(self, "WebhookEndpoint",
+            export_name=f"{stack_name}-{names.WEBHOOK_ENDPOINT}-url",
+            value=endpoint_url("new_recording")
+        )
+
+        core.CfnOutput(self, "OnDemandEndpoint",
+            export_name=f"{stack_name}-{names.ON_DEMAND_ENDPOINT}-url",
+            value=endpoint_url("ingest")
+        )
+
+        core.CfnOutput(self, "WebhookResourceId",
+            export_name=f"{stack_name}-{names.WEBHOOK_ENDPOINT}-resource-id",
+            value=self.new_recording_resource.resource_id
+        )
+
+        core.CfnOutput(self, "OnDemandResourceId",
+            export_name=f"{stack_name}-{names.ON_DEMAND_ENDPOINT}-resource-id",
+            value=self.ingest_resource.resource_id
+        )
+
+        core.CfnOutput(self, "RestApiId",
+            export_name=f"{stack_name}-{names.REST_API}-id",
+            value=self.api.rest_api_id
+        )
```
Putting things like this in cloudformation outputs really makes some of the tasks easier.

nataliemona replied:
So nice!

lbjay commented on cdk/bucket.py:
```diff
@@ -0,0 +1,26 @@
+
+        two_week_lifecycle_rule = s3.LifecycleRule(
+            id="DeleteAfterTwoWeeks",
+            prefix="",  # for development, remove later
```
I don't remember why this is there.

nataliemona replied:
I think this was me... it was for testing something, not needed anymore.

lbjay commented on cdk/codebuild.py:
```diff
@@ -0,0 +1,46 @@
+            ),
+            environment=codebuild.BuildEnvironment(
+                build_image=codebuild.LinuxBuildImage.AMAZON_LINUX_2_2,
+                compute_type=codebuild.ComputeType.MEDIUM
```
i bumped this up from `SMALL` and it only runs about 20% faster. :|

lbjay commented on cdk/function.py:
```diff
@@ -0,0 +1,187 @@
+            "log_retention": logs.RetentionDays.SIX_MONTHS
+        }
+
+        if vpc_id is not None and security_group_id is not None:
+            opencast_vpc = ec2.Vpc.from_lookup(self, "OpencastVpc", vpc_id=vpc_id)
+            opencast_security_group = ec2.SecurityGroup.from_security_group_id(
+                self, "OpencastSecurityGroup", security_group_id=security_group_id
+            )
+            function_props.update({
+                "vpc": opencast_vpc,
+                "security_groups": [opencast_security_group]
+            })
```
I tried to do this only in the uploader function construct but i couldn't get the constructor override, `super().__init__` part to work properly.

lbjay commented on cdk/helpers.py:
```diff
@@ -0,0 +1,74 @@
+from functools import lru_cache
+from functions.common import zoom_api_request
+
+AWS_PROFILE = _getenv("AWS_PROFILE")
+if AWS_PROFILE is not None:
+    boto3.setup_default_session(profile_name=AWS_PROFILE)
+
+# wrap the default getenv so we can enforce required vars
+def getenv(param_name, required=True):
+    val = _getenv(param_name)
+    if required and (val is None or val.strip() == ''):
+        raise Exception(f"Missing environment variable {param_name}")
+    return val
```
this code is repeated in `tasks.py` but if we wanted to consolidate i'm not sure which would be the best home

lbjay replied:
Maybe here since `tasks.py` is already importing `cdk/names.py`

nataliemona replied:
Ok

lbjay commented on cdk/names.py:
```diff
@@ -0,0 +1,42 @@
+
```
so an idea i had for pre-injecting the stack name would be to use a module-level `__getattr__`, Just add to the top top of the module like so:

```
from .helpers import getenv

def __getattr__(attr):
  STACK_NAME =  getenv("STACK_NAME")
  if attr in globals():
    name = globals()[attr]
    return f"{STACK_NAME}-{name}"
  raise AttributeError(f"module {__name__} has no attribute {attr}") 
```

nataliemona replied:
Where?

lbjay commented on cdk/queues.py:
```diff
@@ -0,0 +1,75 @@
+        }
+        for name, queue in queues.items():
+            core.CfnOutput(self, f"{name}-queue-export",
+                export_name=f"{self.stack_name}-{name.replace('.', '-')}-url",
```
cfn exports don't allow '.' in their names, e.g. "foo.fifo"

lbjay commented on cdk/stack.py:
```diff
@@ -0,0 +1,198 @@
+            ]
+        )
+
+        on_demand.add_monitoring(monitoring)
+        webhook.add_monitoring(monitoring)
+        downloader.add_monitoring(monitoring)
+        uploader.add_monitoring(monitoring)
+        op_counts.add_monitoring(monitoring)
+        log_notify.add_monitoring(monitoring)
+        api.add_monitoring(monitoring)
+        queues.add_monitoring(monitoring)
```
I'm imagining adding widgets to a dashboard would work similarly, although our zip dashboard has gotten pretty baroque; it might have to work the opposite way, i.e. the dashboard construct takes the other resource objects as input.

lbjay commented on example.env:
```diff
@@ -64,9 +64,6 @@ MINIMUM_DURATION=2
 # Python pytz timezone
 LOCAL_TIME_ZONE=US/Eastern
 
-# The Lambda Function alias that points to the "live" version of the function
```
I've never once had a reason for this to be configurable so I just made it a constant in `names.py`

lbjay commented on tasks.py:
```diff
@@ -143,11 +106,7 @@ def package(ctx, function=None, upload_to_s3=False):
-    else:
-        functions = FUNCTION_NAMES
-
+    functions = resolve_function_arg(function)
```
I'm really disappointed in the invoke library for not having a way to handle manipulating incoming task arguments. also the `@task` decorator doesn't play nice with other decorators or I would have done it that way. oh well.

lbjay commented on tasks.py:
```diff
@@ -201,114 +156,25 @@ def release(ctx, function=None, description=None):
     req_file = relpath(join(dirname(__file__), 'requirements.in'))
-    ctx.run("pip-compile -r -U {}".format(req_file))
-    for func in FUNCTION_NAMES:
+    ctx.run("pip install -U pip-tools && pip-compile -r -U {}".format(req_file))
```
one time i got some errors because of a `pip-complile` vs `pip` compatibility thing. I think it's good to make sure they're both up-to-date.

nataliemona commented on cdk/function.py:
```diff
@@ -0,0 +1,187 @@
+            "log_retention": logs.RetentionDays.SIX_MONTHS
+        }
+
+        if vpc_id is not None and security_group_id is not None:
```
please can we do `if vpc_id and security_group_id` because this is more concise and also covers other falsy cases such as an empty string

lbjay replied:
:+1:

nataliemona committed:
clean up typos, minor bugs and a few stylistic transgressions

nataliemona commented on cdk/helpers.py:
```diff
@@ -0,0 +1,74 @@
+from functions.common import zoom_api_request
+
+AWS_PROFILE = _getenv("AWS_PROFILE")
+if AWS_PROFILE is not None:
```
`if AWS_PROFILE`... sorry I just started noticing these and they're bothering me

nataliemona commented on cdk/api.py:
```diff
@@ -0,0 +1,188 @@
+
+        log_group = logs.LogGroup(self, "apilogs",
+            log_group_name=f"/aws/apigateway/{self.rest_api_name}/access_logs",
+            removal_policy=core.RemovalPolicy.DESTROY,
```
I added this removal policy - so when you delete the stack it will also delete the logs.

nataliemona commented on .gitignore:
```diff
@@ -6,5 +6,10 @@ __pycache__
 outfile.txt
 .coverage
-venv/
-.gitconfig
```
Good idea. I've started to use my global .gitignore

nataliemona commented on cdk/app.py:
```diff
@@ -0,0 +1,66 @@
+from .stack import ZipStack
+
+STACK_NAME = getenv("STACK_NAME")
+AWS_REGION = getenv("AWS_REGION", required=False) or \
```
I see how this x = "a" or "b" or "c" works now but it wasn't immediately obvious to me. Cool!

nataliemona commented on cdk/function.py:
```diff
@@ -0,0 +1,187 @@
+            )
+            function_props.update({
+                "vpc": opencast_vpc,
+                "security_groups": [opencast_security_group]
```
we're just getting one so `"security_group" : opencast_security_group` is more clear

nataliemona commented on cdk/function.py:
```diff
@@ -0,0 +1,187 @@
+            filter_pattern=logs.FilterPattern.all(
+                logs.JsonPattern(
+                    "$.message.payload.status = \"RECORDING_MEETING_COMPLETED\""
+                )
```
Let's not worry about this now but I think we may not need (any?) metric filters now that we have cloudwatch insights.

nataliemona commented on cdk/helpers.py:
```diff
@@ -0,0 +1,74 @@
+# wrap the default getenv so we can enforce required vars
+def getenv(param_name, required=True):
+    val = _getenv(param_name)
+    if required and (val is None or val.strip() == ''):
```
I think this could be `if required and val`

nataliemona commented on cdk/helpers.py:
```diff
@@ -0,0 +1,74 @@
+            }
+        ]
+    )
+    if "Reservations" not in result or len(result["Reservations"]) == 0:
```
this should be `if "Reservations" not in result or not result["Reservations"]`

nataliemona commented on cdk/queues.py:
```diff
@@ -0,0 +1,75 @@
+            queue_name=f"{self.stack_name}-{names.UPLOAD_QUEUE}",
+            retention_period=core.Duration.days(14),
+            visibility_timeout=core.Duration.seconds(300),
+            content_based_deduplication=False,
```
Why false? Currently in production we have content based deduplication enabled and it protects against duplicate messages due to network issues. It does not prevent requests for the same Zoom recording from being put through during testing because these will have different correlation ids.

nataliemona replied:
Update. You need content based deduplication. If you do not have it, you get this error: `[ERROR] ClientError: An error occurred (InvalidParameterValue) when calling the SendMessage operation: The queue should either have ContentBasedDeduplication enabled or MessageDeduplicationId provided explicitly`

lbjay replied:
:+1:

nataliemona commented on cdk/queues.py:
```diff
@@ -0,0 +1,75 @@
+        # note: you can't simply use the `queue_name` attribute of the
+        # construct objects here: cdk/cfn will complain about
+        # 'The Name field of Export must not depend on any resources'
+        queues = {
```
Yes. I think because they aren't technically created yet or something like that.

nataliemona commented on tasks.py:
```diff
@@ -1694,27 +1188,56 @@ def __request_ids_from_logs(ctx, log_group, filter_pattern):
-}
-"""
+def resolve_function_arg(func):
+    if func is not None:
```
`if func`

nataliemona commented:
I can't approve this PR because it's my PR, but... I approve.

nataliemona commented on cdk/stack.py:
```diff
@@ -0,0 +1,198 @@
+
+        # grant downloader function permissions
+        queues.download_queue.grant_consume_messages(downloader.function)
+        queues.upload_queue.grant_consume_messages(downloader.function)
```
This should be `grant_send_messages`, not having that permission is a blocking bug.

nataliemona commented on cdk/stack.py:
```diff
@@ -0,0 +1,198 @@
+        op_counts.function.grant_invoke(uploader.function)
+        # this is required so that the presigned s3 urls generated by
+        # the uploader function will be valid
+        recordings_bucket.bucket.grant_read(uploader.function)
```
since ffprobe check has been fixed so that it is actually called, we need `recordings_bucket.bucket.grant_read(downloader.function)` here

nataliemona committed:
fix major bugs

nataliemona commented:
Ok. Took my time doing this review but I'm glad I did. It's really important to try at least one test ingest before finalizing code.

I found 3 major issues that would prevent anything from being ingested:
- Downloader lambda did not have permission to read the s3 recording files bucket - this means that the presigned url used to run the ffprobe check is invalid and all ffprobe checks would have failed.
- Instead of permission to send uploader messages, the downloader lambda had permission to consume uploader messages, this meant that everything would fail when the downloader tried to send a message to the uploader.
- The uploader queue had content based deduplication set to false - since fifo queues are required to either have content deduplication or a deduplication id, every ingest would fail.

I've fixed all the major blocking issues, I haven't fixed any of the more minor things that commented on.

I approve. (I can't approve this PR in github officially because technically this is still my own PR).

lbjay committed:
package updates

lbjay committed:
add `stack.*` tasks to wrap the cdk commands

also make function packaging/uploading to s3 part of the stack.create
operation

lbjay committed:
stuff related to aws profiles:

* make sure settings in `.env` override existing env vars.
  note this means you can't, for instance, override `AWS_PROFILE`
  on the command-line if you've set it in `.env`. I *think* this
  is what we want.
* warn if the `cdk` commands are executed directly

lbjay committed:
readme updates related to cdk work

lbjay commented on tasks.py:
```diff
@@ -138,16 +101,63 @@ def codebuild(ctx, revision):
+    """
+    Deletes the cdk CloudFormation stack
+    """
+    ctx.run(f"cdk destroy -c VIA_INVOKE=true {profile_arg()}")
```
this needs to be preceded by emptying the bucket. Even with the retention policy set to `DESTROY` cloudformation won't delete a non-empty bucket: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-s3-bucket.html

lbjay committed:
empty out a bunch of s3 stuff pre stack delete & better delete confirm

lbjay committed:
use the runner's `pty` option

without this the invoke runner can't handle the confirmation prompts
shown by the cdk deployment operation. i waffled a bit on whether we
actually needed to show those.

lbjay committed:
event name could exceed max of 64 char (and was rendundant)

lbjay committed:
a few readme additions/fixes

lbjay committed:
oops: op-counts function needs a db url to connect to

nataliemona committed:
increase downloader lambda memory size

nataliemona committed:
fix redundant naming in function alarms

lbjay committed:
op-counts function needs to get into the opencast vpc network too
